Added Support for Degree Rules

aidengeorge97 opened the pull request:
Degree rules have been mapped as a k-ary tree. See the [database details document](https://docs.google.com/document/d/13rDT9u6-vEgFtUdmtDeBNE_8HUAiiQryGiRG5CbEsxA/edit) or the class docstrings for information on how this is implemented. 

Integrating this pull will allow you to build expressions for DegreeRules and store them on the database. See issue #65 

Sample expression tree representation:


        "id": 3,
        "code": "BARTS",
        "year": 2019,
        "name": "Bachelor of Arts",
        "units": 144,
        "degreeType": "ugrad-sing",
        "rules": [
            "AND",
            [
                "OR",
                [
                    {
                        "code": "ANCH-MAJ",
                        "type": "subplan selection"
                    }
                ],
                [
                    {
                        "code": "HUEB-MAJ",
                        "type": "subplan selection"
                    }
                ],
                [
                    {
                        "code": "TNSO-MAJ",
                        "type": "subplan selection"
                    }
                ]
            ],
            [
                "OR",
                [
                    {
                        "code": "AUIS-MIN",
                        "type": "subplan selection"
                    }
                ],
                [
                    {
                        "code": "DEMO-MIN",
                        "type": "subplan selection"
                    }
                ],
                [
                    {
                        "code": "FORA-MIN",
                        "type": "subplan selection"
                    }
                ]
            ],
            [
                {
                    "type": "electives",
                    "units": 48
                }
            ]
        ]

Commit:
Defined a baseline class for classifying degree rules in JSON format.

Commit:
Refactored DegreeRules to be stored as a k-ary tree and filled out entry functions for various rules.

pep8speaks commented:
Hello @aidengeorge97! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`cassdegrees/api/degreeRules.py`](https://github.com/cass-degrees/CASS-Degrees-Code/blob/7a6923f749b920de44437095accd6a68d17a42d4/cassdegrees/api/degreeRules.py):

> [Line 3:1](https://github.com/cass-degrees/CASS-Degrees-Code/blob/7a6923f749b920de44437095accd6a68d17a42d4/cassdegrees/api/degreeRules.py#L3): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 19:5](https://github.com/cass-degrees/CASS-Degrees-Code/blob/7a6923f749b920de44437095accd6a68d17a42d4/cassdegrees/api/degreeRules.py#L19): [F811](https://duckduckgo.com/?q=pep8%20F811) redefinition of unused '__init__' from line 15
> [Line 161:20](https://github.com/cass-degrees/CASS-Degrees-Code/blob/7a6923f749b920de44437095accd6a68d17a42d4/cassdegrees/api/degreeRules.py#L161): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

JackSekoranja commented on cassdegrees/api/degreeRules.py:
```diff
@@ -1,52 +1,161 @@
+
+        :param expression_list: (list) a list of tokens in the format of a boolean expression.
+                                A token must be one of:
+                                    "(", ")", "AND", "OR", rule
```
Just noting that it looks like this function requires the external tokens to be '(' and ')' as it will not work without them

aidengeorge97 replied:
Yes and no, but I found/fixed the problem you were pointing to anyway.

JackSekoranja commented on cassdegrees/api/degreeRules.py:
```diff
@@ -1,52 +1,161 @@
+        else:
+            self.expressionTree = ["AND", self.expressionTree, rule]
+
+    def get_rules_as_json(self):
```
When generating rules for myself, the first rule was always `null`. Doesn't look intentional

aidengeorge97 replied:
Please elaborate? What were you doing to get a null?

JackSekoranja replied:
@aidengeorge97 This is what I did to get `null`:
```
>>> from api.degreeRules import DegreeRules
>>> d_rules = DegreeRules()
>>> d_rules.build_expression([DegreeRules.rule_electives(24)])
>>> print(d_rules.get_rules_as_json())
[null, [{"type": "electives", "units": 24}]]
```

The same thing also happens with more complex rules such as `d_rules.build_expression(['(', DegreeRules.rule_electives(24), 'AND', DegreeRules.rule_electives(16), ')'])`

aidengeorge97 replied:
Ahhhhhh okay I didn't account for a scenario where a degree would only have one rule. Cause no degrees have one rule. But, valid, I'll just make it so one rule expressions default to an AND so it's a single term AND (i.e. just the one rule).

Not as certain about your second case though, I'll look into it.

aidengeorge97 replied:
Okay I solved the issue. Because you had brackets around those terms, it was expecting another set of rules, hence making a null parent node. I've added a function to flatten the tree, so those kinds of null nodes will be removed and their children get moved up. The logic was still right, but this will neaten the trees up and account for a wider range of poorly designed expressions. 

I also noticed an "issue" with append_rule while changing that, which is now fixed. Rather than just appending singular rules, you can now append a full expression list.

JackSekoranja commented on cassdegrees/api/degreeRules.py:
```diff
@@ -1,52 +1,161 @@
+        :return: (Dict) an elective rule
+        """
+
+        rule = {"type": "electives", "units": units}
```
This is fine for free electives, but there are no rules that allow you to generate list electives (24 units from courses A, B, C, D, E). We will need to have this in order to build more complex degrees

aidengeorge97 replied:
Yep, it's an incomplete set of rules. This is just the base functionality for storing rules and I fully expect many rules to be added as needed.

JackSekoranja commented on cassdegrees/api/degreeRules.py:
```diff
@@ -1,52 +1,161 @@
+        :return: (Dict) a subject requirement rule
+        """
+
+        rule = {"type": "subject requirement", "subject": subject, "level": level, "minimum": minimum, "units": units}
```
This is useful for things like "No more than 24 units of 1000-level ARTS courses", but we may also want to represent restrictions/enforcements on the number of ARTS courses exclusively, as well as enforcements/restrictions on the number of year/level-specific courses exclusively. Function below has the same issue

aidengeorge97 replied:
Valid. Do you want me to finalise all the degree rules before pulling this then? Because that's a full meeting with Deborah and extra requirement gathering. I want to pull this in now so the base functionality is there. Rules are easy to change after that.

JackSekoranja replied:
I just wasn’t sure if this represented a finalised version. You’re probably right in thinking that we should update as we go so I’ll re-check my previous comments tomorrow (well today I guess) to see if they are fixed and will give you the go-ahead if they are

JackSekoranja reviewed (changes requested):
Looks like we'll still need to make some changes to represent all rules. I'm not sure if the intention is for people to add rules to this list as they need them (which may be a good idea), but right now I think there is a lot missing

Commit:
Fixed a minor error with build_expression where no parent case was provided for rule nodes without an existing parent. Added exceptions for common errors and clarified inputs for ease of use.

Commit:
Added support for edge-case degrees which consist of only a single rule.

Commit:
append_rule now takes an expression_list as input, rather than a single rule dict. This allows more complex rules to be appended to existing rules.

Added a flatten_tree function which removes cases where a child has no siblings.

Abstracted build_expression into parse_tree which allows expression trees to be built independent of setting the value of expression_tree.

Commit:
Typo

JackSekoranja reviewed (approved):
Okay, that looks better. Happy for you to merge now
